numpy/core/src/umath/loops.c.src

```diff
@@ -840,6 +846,76 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 void
     UNARY_LOOP_FAST(@type@, @type@, *out = in > 0 ? 1 : (in < 0 ? -1 : 0));
 }
 
+/* Libdivide only supports 32 and 64 bit types
+ * We try to pick the best possible one */
+/**begin repeat1
+ * #kind = t, gen, do#
+ */
+#if NPY_BITSOF_@TYPE@ <= 32
+#define libdivide_@type@_@kind@ libdivide_s32_@kind@
+#else
+#define libdivide_@type@_@kind@ libdivide_s64_@kind@
+#endif
+/**end repeat1**/
+
+#ifndef USE_LEGACY_DIVISION
+NPY_NO_EXPORT void
+@TYPE@_divide(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+    BINARY_LOOP_BASE
+
+    if(!is2) {
+        const @type@ in2 = *(@type@ *)ip2;
+
+        /* Creating a divisor of 0 is treated as an error by libdivide */
+        struct libdivide_@type@_t fast_d = in2 ? libdivide_@type@_gen(in2) : (struct libdivide_@type@_t){0};
+        BINARY_LOOP_FIXED {
+            const @type@ in1 = *(@type@ *)ip1;
+            /*
+             * FIXME: On x86 at least, dividing the smallest representable integer
+             * by -1 causes a SIFGPE (division overflow). We treat this case here
+             * (to avoid a SIGFPE crash at python level), but a good solution would
+             * be to treat integer division problems separately from FPU exceptions
+             * (i.e. a different approach than npy_set_floatstatus_divbyzero()).
+             */
+            if (in2 == 0 || (in1 == NPY_MIN_@TYPE@ && in2 == -1)) {
+                npy_set_floatstatus_divbyzero();
+                *((@type@ *)op1) = 0;
+            }
+            else {
+                *((@type@ *)op1) = libdivide_@type@_do(in1, &fast_d);
+
+                if(((in1 > 0) != (in2 > 0)) && (*((@type@ *)op1) * in2 != in1)) {
```

Comment:
```suggestion
                if (((in1 > 0) != (in2 > 0)) && (*((@type@ *)op1) * in2 != in1)) {
```
Maybe we should add a comment that this rounds down negative reuslts instead of rounding towards 0?